android-sdk/README.md

```diff
@@ -0,0 +1,186 @@
+
+Example KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+   /* snip */
+   public KeyCloakService() {}
+
+   @Override
+    public void bootstrap(MobileCore core, ServiceConfiguration config) {
+        this.serverUrl = config.getProperty("auth-server-url");
+        this.clientId = config.getProperty("client_id");
+        this.audience = config.getProperty("audience");
+        this.grantType = config.getProperty("grant_type");
+        this.subjectTokenType = config.getProperty("subject_token_type");
+        this.requestedTokenType = config.getProperty("requested_token_type");
+        this.realm = config.getProperty("realm");
+        this.core = core;
+    }
+}
+```
+
+#### Registering a Service Module and the ServiceModuleRegistry
+The Core SDK will provide a service registry that modules can register themselves with as well as register their dependencies.  During build MobileCore will call bootstrap on services in the order of those with the least dependencies to those with the most dependencies.
+
+Example registration of KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+  static {
+    ServiceModuleRegistry.registerServiceModule("keycloak", KeyCloakService.class, "http");
+  }
+  /*snip*/
+}
+```
+
+ServiceModuleRegistry.java
+```
+
+/**
+ * Registry of service modules.  It is responsible for maintaining a mapping of service modules to
+ * their types as well as the dependencies each module expects.
+ */
+public final class ServiceModuleRegistry {
+    private static final Map<String, Class<? extends ServiceModule>> serviceTypeMap = new HashMap<>();
+    private static final Map<String, String> dependencyMap = new HashMap<>();
+    /**
+     * Register the looking for a service module to a class.
+     *
+     * @param type the type of module to use
+     * @param moduleClass the class that implements a service module type
+     * @param dependsOn the other serviceTypes this Service Depends on.
+     */
+    public static void registerServiceModule(String type, Class<? extends ServiceModule> moduleClass, String... dependsOn) {
+        serviceTypeMap.put(type, moduleClass);
+        for (String dependency : dependsOn) {
+            dependencyMap.put(type, dependency);
+        }
+    }
+
+}
+```
+** Discussion Point ** I'm not a fan of static methods during testing.  Perhaps there should be a method in the ServiceModule interface which returns a list of services that the service depends on.  MobileCore could then read that list by invoking that method as part of bootstrapping.
+
+#### Referencing a Service
+An instance of the core bassed during bootstrap can be used to reference other services.  MobileCore is responsible for ensuring that ServiceModule.bootstrap is not called until all declared dependencies are bootstrapped.
```

Comment:
(I'm assuming you're using MobileCore and Core SDK interchangeably here)
How will MobileCore know when this is? 
Does the developer have to do that at a point in their code where all services have been declared/registered with the Core SDK?

Reply:
MobileCore is a class. Core SDK is the SDK.  Any places where it is ambiguous should be clarified.

MobileCore will be resposible for bootstrapping methods during instantiation/initialization. (see the https://github.com/secondsun/aerogear-android-sdk/blob/87b4ff629b6f3c47040d2e322277e7cee3bcb76f/core/src/test/java/org/aerogear/mobile/core/MobileCoreParserTest.java#L25)

Services have to be declared in the service registry before then.

Reply:
So is the instantiation/initialization something the developer triggers programatically in their app code? e.g. `mobileCore.initialize()`
And they have to make sure they've declared all services with the registry before they call that method?

Reply:
@david-martin Has really good point. I think that proposal should first define top level user api. For example:

1. Add core and serviceA dependency to gradle
2. Create Core instance by.. etc.

This will help us to see how users will use it and we can then describe internal parts

Reply:
I have done example: https://gist.github.com/wtrocki/5774f883a9a2ae0b33c5aa63986cae45